### PR TITLE
Constrained monomorphs

### LINKED ISSUE #9393: Constraint check error position is not accurate
In the example above, the "constraint check failure" error will be reported at the position of `f` (not even the call). I think I understand why that's happening (it's checking the field itself, not really a call argument, right?), but maybe it is possible to store a position of a thing that made T an actual type (`1` in this example).

```haxe
class Main {
	static function f<T:String>(v:T) {}
	static function main() {
		f(1);
	}
}
```

### PR DESCRIPTION
**Pull request by Simn**
This PR adds support for constraints on monomorphs by introducing the `tm_constraints` field on `tmono`. When a monomorph is bound via `Monomorph.bind`, the compiler checks if the binding is valid in `check_constraints`.

There are currently two sources of constraints:

1. *Type parameters with constraints*: These now all go through `spawn_constrained_monos`, which creates the monomorph and adds the constraints from type parameters, if exists.
2. *Field access*: This replaces open anonymous structures. When a field is accessed on a monomorph that is either unconstrained or has the `MOpenStructure` flag, a structural constraint for that field is added to the monomorph.

Furthermore, if we have an unbound monomorph after typing a function, that monomorph is `Monomorph.close`d by using its constraints to find a type.

Some information about related issues:

* Because constraints are now checked upon binding, we have more context. This fixes #9393.
* The example in #5225 works now, but there's more work to be done here.
* The lua and JVM generators run into problems with #8799. This happens because we might now get more accurate type inference, so an anonymous field access might become a proper class field access now.

For now, my goal here is to "not break anything". Afterwards, we can look into what further possibilities this opens. This includes:

* An alternative to `@:multiType`
* Removal of the `Opened` flag for structures
* Additional constraints like array access

I'm not 100% confident in the changes made here, but we have to start somewhere.

### CODE REVIEW

File: src/core/tUnification.ml [hunk 1]
```diff
@@ -53,14 +53,122 @@ let default_unification_context = {
...
+		m.tm_type <- Some t;
+		m.tm_constraints <- []
 
 	let rec bind m t =
-		m.tm_type <- Some t
+		begin match t with
+		| TAnon _ when List.mem MOpenStructure m.tm_constraints ->
+			(* If we assign an open structure monomorph to another structure, the semantics want us to merge the
+			   fields. This is kinda weird, but that's how it has always worked. *)
+			constrain_to_type m None t;
```
Comment (RealyUniqueName) [inline src/core/tUnification.ml:130]: Shouldn't we also check if anon's fields don't conflict with current constraints here?
I didn't find such a check down the `constraint_to_type` calls path, and `Monomorph.bind` is called from various places other than unification.

Comment (Simn) [reply src/core/tUnification.ml:130]: Yes I think a check is missing here. I think we can do this in `classify_constraints`, where we're building the fields map anyway. This should avoid any problems with too early unification.

Comment (RealyUniqueName) [reply src/core/tUnification.ml:130]: Same question for other branches of this `match`

File: src/context/typecore.ml [hunk 2]
```diff
@@ -519,13 +519,43 @@ let merge_core_doc ctx mt =
...
 let spawn_constrained_monos ctx p map params =
 	let monos = List.map (fun (s,_) ->
 		let mono = Monomorph.create() in
 		TMono mono
 	) params in
+	let map t = map (apply_params params monos t) in
+	check_constraints map params monos p;
 	monos
 
+let safe_mono_close ctx m p =
```
Comment (RealyUniqueName) [inline src/context/typecore.ml:551]: Return value of this function is never used. 
It should be changed to `unit` probably. And `Monomorph.close` too.

**Comment by RealyUniqueName** [comment]
> I'm not 100% confident in the changes made here, but we have to start somewhere.

I 100% agree. Let's merge this and deal with the issues should they arise.

**Comment by Simn** [comment]
I'll deal with field constraint conflicts later. That requires some proper handling to avoid getting `Unify_error` exceptions thrown in places we don't want them. The problem seems pretty difficult to reproduce anyway.

**Pull request merged**